File: shared/serverapi/task_search.go
package serverapi

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"unicode/utf8"

	"core/shared/protocol"
	"core/shared/tasksearchtext"
)

const (
	TaskSearchMaxQueryRunes   = 4096
	TaskSearchDefaultContext  = 20
	TaskSearchMinContext      = 1
	TaskSearchMaxContext      = 64
	TaskSearchDefaultPageSize = 100
	TaskSearchMaxPageSize     = 100

	TaskSearchSparseDocumentContractVersion = "kent-task-search-sparse-document-v1"
	TaskSearchRankingContractVersion        = "kent-task-search-ranking-v1"
)

type TaskSearchMode string

const (
	TaskSearchModeLiteral TaskSearchMode = "literal"
	TaskSearchModeFTS5    TaskSearchMode = "fts5"
)

type TaskSearchSourceKind string

const (
	TaskSearchSourceKindShortID TaskSearchSourceKind = "short_id"
	TaskSearchSourceKindTitle   TaskSearchSourceKind = "title"
	TaskSearchSourceKindBody    TaskSearchSourceKind = "body"
	TaskSearchSourceKindComment TaskSearchSourceKind = "comment"
)

type TaskSearchRequest struct {
	Mode            TaskSearchMode           `json:"mode"`
	Query           string                   `json:"query"`
	Context         int                      `json:"context"`
	CaseSensitive   bool                     `json:"case_sensitive"`
	IncludeComments bool                     `json:"include_comments"`
	ProjectIDs      []string                 `json:"project_ids,omitempty"`
	StatusKinds     []WorkflowTaskStatusKind `json:"status_kinds,omitempty"`
	PageSize        int                      `json:"page_size"`
	Offset          *int                     `json:"offset,omitempty"`
}

func (r TaskSearchRequest) Validate() error {
	if r.Mode != TaskSearchModeLiteral && r.Mode != TaskSearchModeFTS5 {
		return taskSearchFieldError("mode", "mode is invalid")
	}
	if r.Query == "" || strings.TrimSpace(r.Query) == "" {
		return taskSearchFieldError("query", "query is required")
	}
	if strings.TrimSpace(r.Query) != r.Query {
		return taskSearchFieldError("query", "query must be trimmed")
	}
	if utf8.RuneCountInString(r.Query) > TaskSearchMaxQueryRunes {
		return taskSearchFieldError("query", "query is too long")
	}
	if r.Mode == TaskSearchModeLiteral && tasksearchtext.NormalizedLiteralRuneCount(r.Query) < 3 {
		return &TaskSearchError{Reason: TaskSearchErrorReasonNormalizedTooShort}
	}
	if r.Mode == TaskSearchModeFTS5 && r.CaseSensitive {
		return taskSearchFieldError("case_sensitive", "case_sensitive requires literal mode")
	}
	if r.Context < TaskSearchMinContext || r.Context > TaskSearchMaxContext {
		return taskSearchFieldError("context", "context is out of range")
	}
	if r.PageSize < 1 || r.PageSize > TaskSearchMaxPageSize {
		return taskSearchFieldError("page_size", "page_size is out of range")
	}
	if r.Offset != nil && *r.Offset < 0 {
		return taskSearchFieldError("offset", "offset must be non-negative")
	}
	for index, projectID := range r.ProjectIDs {
		if strings.TrimSpace(projectID) == "" || strings.TrimSpace(projectID) != projectID {
			return taskSearchFieldError(fmt.Sprintf("project_ids[%d]", index), "project id is invalid")
		}
		if index > 0 && r.ProjectIDs[index-1] >= projectID {
			return taskSearchFieldError("project_ids", "project ids must be sorted and unique")
		}
	}
	for index, status := range r.StatusKinds {
		if _, valid := status.NativeState(); !valid {
			return taskSearchFieldError(fmt.Sprintf("status_kinds[%d]", index), "status kind is invalid")
		}
		if index > 0 && r.StatusKinds[index-1] >= status {
			return taskSearchFieldError("status_kinds", "status kinds must be sorted and unique")
		}
	}
	return nil
}

type TaskSearchResponse struct {
	Mode       TaskSearchMode    `json:"mode"`
	Groups     []TaskSearchGroup `json:"groups"`
	NextOffset *int              `json:"next_offset,omitempty"`
}

type TaskSearchGroup struct {
	ProjectID     string             `json:"project_id"`
	ProjectKey    string             `json:"project_key"`
	TaskID        string             `json:"task_id"`
	ShortID       string             `json:"short_id"`
	WorkflowID    string             `json:"workflow_id"`
	Title         string             `json:"title"`
	Status        WorkflowTaskStatus `json:"status"`
	TotalHitCount int                `json:"total_hit_count"`
	Hits          []TaskSearchHit    `json:"hits"`
}

type TaskSearchHit struct {
	Ordinal int                   `json:"ordinal"`
	Source  TaskSearchSource      `json:"source"`
	Literal *TaskSearchLiteralHit `json:"literal,omitempty"`
	FTS5    *TaskSearchFTS5Hit    `json:"fts5,omitempty"`
}

type TaskSearchSource struct {
	Kind      TaskSearchSourceKind `json:"kind"`
	CommentID *string              `json:"comment_id,omitempty"`
}

type TaskSearchLiteralHit struct {
	Before         string `json:"before"`
	Match          string `json:"match"`
	After          string `json:"after"`
	LeftTruncated  bool   `json:"left_truncated"`
	RightTruncated bool   `json:"right_truncated"`
}

type TaskSearchFTS5Hit struct {
	Snippet string `json:"snippet"`
}

func (r TaskSearchResponse) Validate() error {
	if r.Mode != TaskSearchModeLiteral && r.Mode != TaskSearchModeFTS5 {
		return errors.New("task search response mode is invalid")
	}
	if r.Groups == nil {
		return errors.New("task search response groups are required")
	}
	if r.NextOffset != nil && *r.NextOffset <= 0 {
		return errors.New("task search response next offset is invalid")
	}
	groupTaskIDs := make(map[string]struct{}, len(r.Groups))
	for groupIndex, group := range r.Groups {
		for _, field := range []struct {
			name  string
			value string
		}{
			{name: "project id", value: group.ProjectID},
			{name: "project key", value: group.ProjectKey},
			{name: "task id", value: group.TaskID},
			{name: "short id", value: group.ShortID},
			{name: "workflow id", value: group.WorkflowID},
			{name: "title", value: group.Title},
		} {
			if err := validateTaskSearchResponseString(field.name, field.value); err != nil {
				return fmt.Errorf("task search group %d: %w", groupIndex, err)
			}
		}
		if _, exists := groupTaskIDs[group.TaskID]; exists {
			return fmt.Errorf("task search response duplicates task group %q", group.TaskID)
		}
		groupTaskIDs[group.TaskID] = struct{}{}
		if len(group.Hits) == 0 || group.TotalHitCount < len(group.Hits) || group.TotalHitCount < 1 {
			return fmt.Errorf("task search group %d total hit count is invalid", groupIndex)
		}
		if err := validateTaskSearchStatus(group.Status); err != nil {
			return fmt.Errorf("task search group %d: %w", groupIndex, err)
		}
		for hitIndex, hit := range group.Hits {
			if err := hit.Validate(r.Mode); err != nil {
				return err
			}
			if hit.Ordinal > group.TotalHitCount {
				return fmt.Errorf("task search group %d hit %d ordinal exceeds total hit count", groupIndex, hitIndex)
			}
			if hitIndex > 0 && group.Hits[hitIndex-1].Ordinal >= hit.Ordinal {
				return fmt.Errorf("task search group %d hit ordinals are not strictly ascending", groupIndex)
			}
		}
	}
	return nil
}

func (h TaskSearchHit) Validate(mode TaskSearchMode) error {
	if h.Ordinal < 1 {
		return errors.New("task search hit ordinal must be positive")
	}
	switch h.Source.Kind {
	case TaskSearchSourceKindShortID:
		if mode != TaskSearchModeLiteral {
			return errors.New("task search Short ID source requires literal mode")
		}
		if h.Source.CommentID != nil {
			return errors.New("task search Short ID source forbids comment id")
		}
	case TaskSearchSourceKindTitle, TaskSearchSourceKindBody:
		if h.Source.CommentID != nil {
			return errors.New("task search title/body source forbids comment id")
		}
	case TaskSearchSourceKindComment:
		if h.Source.CommentID == nil {
			return errors.New("task search comment source requires comment id")
		}
		if err := validateTaskSearchResponseString("comment id", *h.Source.CommentID); err != nil {
			return err
		}
	default:
		return errors.New("task search source kind is invalid")
	}
	if mode == TaskSearchModeLiteral && h.Literal != nil && h.FTS5 == nil {
		if h.Literal.Match == "" {
			return errors.New("task search literal hit match is required")
		}
		return nil
	}
	if mode == TaskSearchModeFTS5 && h.FTS5 != nil && h.Literal == nil {
		if h.FTS5.Snippet == "" {
			return errors.New("task search FTS5 hit snippet is required")
		}
		return nil
	}
	return errors.New("task search hit mode payload is invalid")
}

type TaskSearchErrorReason string

const (
	TaskSearchErrorReasonNormalizedTooShort TaskSearchErrorReason = "normalized_too_short"
)

type TaskSearchError struct {
	Reason TaskSearchErrorReason `json:"reason"`
}

func (e *TaskSearchError) Error() string {
	if e == nil {
		return "task search error"
	}
	return "task search error: " + string(e.Reason)
}

func (e TaskSearchError) Validate() error {
	switch e.Reason {
	case TaskSearchErrorReasonNormalizedTooShort:
		return nil
	default:
		return errors.New("task search error reason is invalid")
	}
}

func (e *TaskSearchError) RPCErrorCode() int {
	return protocol.ErrCodeWorkflowTaskSearch
}

func (e *TaskSearchError) RPCErrorData() json.RawMessage {
	if e == nil || e.Validate() != nil {
		return nil
	}
	return marshalRPCErrorData(struct {
		Type   string                `json:"type"`
		Reason TaskSearchErrorReason `json:"reason"`
	}{
		Type:   "task_search_error",
		Reason: e.Reason,
	})
}

func DecodeTaskSearchError(data json.RawMessage, message string) error {
	var envelope struct {
		Type   string                `json:"type"`
		Reason TaskSearchErrorReason `json:"reason"`
	}
	if err := json.Unmarshal(data, &envelope); err != nil || envelope.Type != "task_search_error" {
		return taskSearchFallbackError(message)
	}
	decoded := &TaskSearchError{Reason: envelope.Reason}
	if err := decoded.Validate(); err != nil {
		return taskSearchFallbackError(message)
	}
	return decoded
}

func taskSearchFieldError(field string, message string) error {
	return WorkflowRequestValidationError{Code: WorkflowRequestErrorInvalidValue, Field: field, Message: message}
}

func taskSearchFallbackError(message string) error {
	if trimmed := strings.TrimSpace(message); trimmed != "" {
		return errors.New(trimmed)
	}
	return errors.New("task search error")
}

func validateTaskSearchResponseString(field string, value string) error {
	if strings.TrimSpace(value) == "" || strings.TrimSpace(value) != value {
		return fmt.Errorf("task search response %s is invalid", field)
	}
	return nil
}

func validateTaskSearchStatus(status WorkflowTaskStatus) error {
	nativeState, valid := status.Kind.NativeState()
	if !valid || status.NativeState != nativeState {
		return errors.New("task search response status is invalid")
	}
	return nil
}

File: apps/desktop/src/api/schemas/taskSearch.ts
import { z } from "zod";

import type { TaskStatusKind } from "../models";
import type {
  TaskSearchFTS5Hit,
  TaskSearchGroup,
  TaskSearchLiteralHit,
  TaskSearchResponse,
  TaskSearchSource,
} from "../taskSearch";
import { taskStatusSchema } from "./common";

const exactNonBlankString = z
  .string()
  .min(1)
  .refine((value) => value.trim() === value);

const nativeStateByKind = {
  done: "terminal",
  waiting_question: "waiting_ask",
  waiting_approval: "waiting_approval",
  interrupted: "interrupted",
  running: "running",
  queued: "queued",
  backlog: "active",
  active: "active",
} as const satisfies Readonly<Record<TaskStatusKind, string>>;

const taskSearchStatusSchema = taskStatusSchema.superRefine((value, context) => {
  if (value.nativeState !== nativeStateByKind[value.kind]) {
    context.addIssue({
      code: "custom",
      message: "native state does not match kind",
      path: ["nativeState"],
    });
  }
});

const titleOrBodySourceSchema = z
  .object({ kind: z.enum(["title", "body"]) })
  .strict()
  .transform((value): TaskSearchSource => value);

const shortIDSourceSchema = z
  .object({ kind: z.literal("short_id") })
  .strict()
  .transform((value): TaskSearchSource => value);

const commentSourceSchema = z
  .object({
    kind: z.literal("comment"),
    comment_id: exactNonBlankString,
  })
  .strict()
  .transform((value): TaskSearchSource => ({
    kind: value.kind,
    commentID: value.comment_id,
  }));

const literalSourceSchema = z.union([
  shortIDSourceSchema,
  titleOrBodySourceSchema,
  commentSourceSchema,
]);
const fts5SourceSchema = z.union([titleOrBodySourceSchema, commentSourceSchema]);

const literalHitSchema: z.ZodType<TaskSearchLiteralHit> = z
  .object({
    ordinal: z.number().int().positive(),
    source: literalSourceSchema,
    literal: z
      .object({
        before: z.string(),
        match: z.string().min(1),
        after: z.string(),
        left_truncated: z.boolean(),
        right_truncated: z.boolean(),
      })
      .strict(),
  })
  .strict()
  .transform((value) => ({
    ordinal: value.ordinal,
    source: value.source,
    literal: {
      before: value.literal.before,
      match: value.literal.match,
      after: value.literal.after,
      leftTruncated: value.literal.left_truncated,
      rightTruncated: value.literal.right_truncated,
    },
  }));

const fts5HitSchema: z.ZodType<TaskSearchFTS5Hit> = z
  .object({
    ordinal: z.number().int().positive(),
    source: fts5SourceSchema,
    fts5: z.object({ snippet: z.string().min(1) }).strict(),
  })
  .strict();

function groupSchema(
  hitSchema: z.ZodType<TaskSearchLiteralHit> | z.ZodType<TaskSearchFTS5Hit>,
): z.ZodType<TaskSearchGroup> {
  return z
    .object({
      project_id: exactNonBlankString,
      project_key: exactNonBlankString,
      task_id: exactNonBlankString,
      short_id: exactNonBlankString,
      workflow_id: exactNonBlankString,
      title: exactNonBlankString,
      status: taskSearchStatusSchema,
      total_hit_count: z.number().int().positive(),
      hits: z.array(hitSchema).min(1),
    })
    .strict()
    .superRefine((value, context) => {
      if (value.total_hit_count < value.hits.length) {
        context.addIssue({
          code: "custom",
          message: "total_hit_count is smaller than returned hits",
          path: ["total_hit_count"],
        });
      }
      let previousOrdinal = 0;
      for (const [index, hit] of value.hits.entries()) {
        if (hit.ordinal <= previousOrdinal || hit.ordinal > value.total_hit_count) {
          context.addIssue({
            code: "custom",
            message: "hit ordinals must be strictly ascending and in range",
            path: ["hits", index, "ordinal"],
          });
        }
        previousOrdinal = hit.ordinal;
      }
    })
    .transform((value) => ({
      projectID: value.project_id,
      projectKey: value.project_key,
      taskID: value.task_id,
      shortID: value.short_id,
      workflowID: value.workflow_id,
      title: value.title,
      status: value.status,
      totalHitCount: value.total_hit_count,
      hits: value.hits,
    }));
}

function responseSchema(
  mode: "literal" | "fts5",
  hitSchema: z.ZodType<TaskSearchLiteralHit> | z.ZodType<TaskSearchFTS5Hit>,
): z.ZodType<TaskSearchResponse> {
  return z
    .object({
      mode: z.literal(mode),
      groups: z.array(groupSchema(hitSchema)),
      next_offset: z.number().int().positive().nullish(),
    })
    .strict()
    .superRefine((value, context) => {
      const taskIDs = new Set<string>();
      for (const [index, group] of value.groups.entries()) {
        if (taskIDs.has(group.taskID)) {
          context.addIssue({
            code: "custom",
            message: "task groups must be unique",
            path: ["groups", index, "task_id"],
          });
        }
        taskIDs.add(group.taskID);
      }
    })
    .transform((value) => ({
      mode: value.mode,
      groups: value.groups,
      nextOffset: value.next_offset ?? null,
    }));
}

export const taskSearchResponseSchema: z.ZodType<TaskSearchResponse> = z.union([
  responseSchema("literal", literalHitSchema),
  responseSchema("fts5", fts5HitSchema),
]);
